File: commons-core/src/main/java/xyz/wagyourtail/commons/core/StringUtils.java
package xyz.wagyourtail.commons.core;

public class StringUtils {

    private StringUtils() {
    }

    public static int count(String str, char c) {
        int count = 0;
        for (char ch : str.toCharArray()) {
            if (ch == c) {
                count++;
            }
        }
        return count;
    }

    public static String translateEscapes(String str) {
        return translateEscapes(str, false);
    }

    public static String translateEscapes(String str, boolean lenient) {
        if (str.isEmpty() || !str.contains("\\")) {
            return str;
        }
        StringBuilder sb = new StringBuilder(str.length());
        int i = 0;
        while (i < str.length()) {
            char c = str.charAt(i++);
            if (c == '\\') {
                if (i >= str.length()) throw new IllegalArgumentException("Invalid escape, hit end of string");
                char n = str.charAt(i++);
                switch (n) {
                    case '"':
                        sb.append('"');
                        break;
                    case '\'':
                        sb.append('\'');
                        break;
                    case '\\':
                        sb.append('\\');
                        break;
                    case '0':
                    case '1':
                    case '2':
                    case '3':
                    case '4':
                    case '5':
                    case '6':
                    case '7':
                        int max;
                        if (n < '4') {
                            max = 2;
                        } else {
                            max = 1;
                        }
                        StringBuilder octal = new StringBuilder();
                        octal.append(n);
                        for (int j = 0; j < max; j++) {
                            if (i >= str.length()) break;
                            char next = str.charAt(i);
                            if (next < '0' || next > '7') break;
                            octal.append(next);
                            i++;
                        }
                        sb.append(Character.toChars(Integer.parseInt(octal.toString(), 8)));
                        break;
                    case 'u':
                        if (i + 4 > str.length())
                            throw new IllegalArgumentException("Invalid unicode escape, hit end of string");
                        String hex = str.substring(i, i + 4);
                        sb.append(Character.toChars(Integer.parseInt(hex, 16)));
                        i += 4;
                        break;
                    case 'b':
                        sb.append('\b');
                        break;
                    case 'f':
                        sb.append('\f');
                        break;
                    case 'n':
                        sb.append('\n');
                        break;
                    case 'r':
                        sb.append('\r');
                        break;
                    case 's':
                        sb.append(' ');
                        break;
                    case 't':
                        sb.append('\t');
                        break;
                    default:
                        if (lenient) {
                            sb.append('\\');
                            sb.append(n);
                        } else {
                            throw new IllegalArgumentException("Invalid escape: " + n + " in \"" + str + "\" at " + i);
                        }
                }
            } else {
                sb.append(c);
            }
        }
        return sb.toString();
    }

    public static String escape(String str) {
        return escape(str, false, false, false);
    }

    public static String escape(String str, boolean unicode) {
        return escape(str, unicode, false, false);
    }

    public static String escape(String str, boolean unicode, boolean spaces, boolean doubleQuote) {
        if (str.isEmpty()) return str;
        StringBuilder sb = new StringBuilder(str.length());
        int i = 0;
        while (i < str.length()) {
            char c = str.charAt(i++);
            switch (c) {
                case '"':
                    if (doubleQuote) {
                        sb.append("\"\"");
                    } else {
                        sb.append("\\\"");
                    }
                    break;
                case '\'':
                    sb.append("\\'");
                    break;
                case '\\':
                    sb.append("\\\\");
                    break;
                case '\b':
                    sb.append("\\b");
                    break;
                case '\f':
                    sb.append("\\f");
                    break;
                case '\n':
                    sb.append("\\n");
                    break;
                case '\r':
                    sb.append("\\r");
                    break;
                case '\t':
                    sb.append("\\t");
                    break;
                case ' ':
                    if (spaces) {
                        sb.append("\\s");
                    } else {
                        sb.append(c);
                    }
                    break;
                default:
                    if (unicode && (c < 0x20 || c > 0x7f)) {
                        sb.append("\\u").append(Integer.toHexString(c).toUpperCase());
                    } else {
                        sb.append(c);
                    }
            }
        }
        return sb.toString();
    }

    public static String toHex(byte[] hexBytes) {
        StringBuilder hex = new StringBuilder();
        for (byte b : hexBytes) {
            hex.append(String.format("%02X", b));
        }
        return hex.toString();
    }

    public static String capitalize(String str) {
        if (str.isEmpty()) return str;
        return Character.toUpperCase(str.charAt(0)) + str.substring(1);
    }

    public static String uncapitalize(String str) {
        if (str.isEmpty()) return str;
        return Character.toLowerCase(str.charAt(0)) + str.substring(1);
    }

    public static int indexOf(String str, char c, int startIndex, int endIndex) {
        int index = str.indexOf(c, startIndex);
        if (index >= endIndex) return -1;
        return index;
    }

}

File: commons-kt-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/utils/StringUtils.kt
package xyz.wagyourtail.commonskt.utils

import kotlin.text.indexOf

/**
 * unlike java, this will translate unicode escapes as well
 */
fun String.translateEscapes(lenient: Boolean = false): String {
    if (this.isEmpty() || !this.contains("\\")) {
        return this
    }
    return buildString(this.length) {
        var i = 0
        while (i < this@translateEscapes.length) {
            val c = this@translateEscapes[i++]
            if (c == '\\') {
                if (i >= this@translateEscapes.length) throw IllegalArgumentException("Invalid escape, hit end of string")
                when (val n = this@translateEscapes[i++]) {
                    '"' -> append('"')
                    '\'' -> append('\'')
                    '\\' -> append('\\')
                    '0', '1', '2', '3', '4', '5', '6', '7' -> {
                        var octal = n.toString()
                        val max = if (octal.toInt() < 4) 2 else 1
                        for (j in 0 until max) {
                            if (i >= this@translateEscapes.length) break
                            val next = this@translateEscapes[i]
                            if (next !in ('0'..'7')) break
                            octal += next
                            i++
                        }
                        append(octal.toInt(8).toChar())
                    }

                    'u' -> {
                        val hex = this@translateEscapes.substring(i, i + 4)
                        if (hex.length != 4) throw IllegalArgumentException("Invalid unicode escape: $hex, expected 4 characters, found EOS")
                        append(hex.toInt(16).toUnicode())
                        i += 4
                    }

                    'b' -> append('\b')
                    'f' -> append('\u000C')
                    'n' -> append('\n')
                    'r' -> append('\r')
                    's' -> append(' ')
                    't' -> append('\t')
                    else -> {
                        if (lenient) {
                            append("\\")
                            append(n)
                        } else {
                            throw IllegalArgumentException(
                                "Invalid escape: ${n.toString().escape(true)} in \"${this}\" at $i"
                            )
                        }
                    }
                }
            } else {
                append(c)
            }
        }
    }
}

fun String.escape(unicode: Boolean = false, spaces: Boolean = false, doubleQuote: Boolean = false): String {
    if (this.isEmpty()) return this
    return buildString(this.length) {
        var i = 0
        while (i < this@escape.length) {
            when (val c = this@escape[i++]) {
                '"' -> {
                    if (doubleQuote) {
                        append("\"\"")
                    } else {
                        append("\\\"")
                    }
                }

                '\'' -> append("\\'")
                '\\' -> append("\\\\")
                '\b' -> append("\\b")
                '\u000C' -> append("\\f")
                '\n' -> append("\\n")
                '\r' -> append("\\r")
                '\t' -> append("\\t")
                ' ' -> {
                    if (spaces) {
                        append("\\s")
                    } else {
                        append(c)
                    }
                }

                else -> {
                    if (unicode && (c.code < 0x20 || c.code > 0x7f)) {
                        append("\\u${c.code.toString(16).padStart(4, '0')}")
                    } else {
                        append(c)
                    }
                }
            }
        }
    }
}

fun String.maybeEscape(): String {
    if (any { it.isWhitespace() } || startsWith("\"")) {
        return "\"${escape(unicode = true)}\""
    }
    return this
}

fun String.removeSuffix(suffix: String, ignoreCase: Boolean = false): String {
    if (endsWith(suffix, ignoreCase)) {
        return substring(0, length - suffix.length)
    }
    return this
}

fun String.removePrefix(prefix: String, ignoreCase: Boolean = false): String {
    if (startsWith(prefix, ignoreCase)) {
        return substring(prefix.length)
    }
    return this
}

fun String.removeSurrounding(prefix: String, suffix: String, ignoreCase: Boolean = false): String {
    return removePrefix(prefix, ignoreCase).removeSuffix(suffix, ignoreCase)
}

fun String.capitalized() = replaceFirstChar { if (it.isLowerCase()) it.uppercase() else it.toString() }

fun String.uncapitalized() = replaceFirstChar { if (it.isUpperCase()) it.lowercase() else it.toString() }

fun String.decapitalized() = uncapitalized()

fun String.indexOf(other: Char, startIndex: Int, endIndex: Int, ignoreCase: Boolean = false): Int {
    val i = indexOf(other, startIndex, ignoreCase)
    if (i >= endIndex) return -1
    return i
}

fun String.count(char: Char, startIndex: Int = 0, endIndex: Int = this.length, ignoreCase: Boolean = false): Int {
    var count = 0
    var i = startIndex
    while (i < endIndex) {
        val index = indexOf(char, i, ignoreCase)
        if (index < 0) break
        count++
        i = index + 1
    }
    return count
}

fun String.count(subString: String, startIndex: Int = 0, endIndex: Int = this.length, ignoreCase: Boolean = false): Int {
    var count = 0
    var i = startIndex
    while (i < endIndex) {
        val index = indexOf(subString, i, ignoreCase)
        if (index < 0) break
        count++
        i = index + subString.length
    }
    return count
}